witness/golang/cmd/witness/internal/witness/witness.go

```diff
@@ -121,120 +121,113 @@ func (w *Witness) GetLogs() ([]string, error) {
 	// avoid race conditions when updating the database.
 	tx, err := w.db.BeginTx(ctx, nil)
 	if err != nil {
-		return 0, fmt.Errorf("couldn't create tx: %v", err)
+		return nil, fmt.Errorf("couldn't create db tx: %v", err)
 	}
-	// Get the latest checkpoint and compact range (if one exists).
+	// Get the latest checkpoint (if one exists) and compact range.
 	prevRaw, rangeRaw, err := w.getLatestChkptData(tx.QueryRow, logID)
 	if err != nil {
 		// If there was nothing stored already then treat this new
-		// checkpoint as trust-on-first-use.
+		// checkpoint as trust-on-first-use (TOFU).
 		if status.Code(err) == codes.NotFound {
-			// If we're using compact ranges then store the input
-			// proof as our initial range, assuming it matches the
-			// input checkpoint.
-			if logInfo.UseCompact {
-				rf := compact.RangeFactory{Hash: logInfo.Hasher.HashChildren}
-				rng, err := rf.NewRange(0, next.Size, proof)
-				if err != nil {
-					return 0, fmt.Errorf("can't form compact range: %v", err)
-				}
-				if err := verifyRangeHash(next.Hash, rng); err != nil {
-					return 0, fmt.Errorf("input root hash doesn't verify: %v", err)
-				}
-				r := []byte(log.Proof(proof).Marshal())
-				return 0, w.setChkptData(tx, logID, nextRaw, r)
+			if err := w.setInitChkptData(tx, logID, next, signed, proof); err != nil {
+				return nil, fmt.Errorf("couldn't set TOFU checkpoint: %v", err)
 			}
-			// If we're not using compact ranges no need to store one.
-			return 0, w.setChkptData(tx, logID, nextRaw, nil)
+			return signed, nil
 		}
-		return 0, fmt.Errorf("Update: %w", err)
+		return nil, fmt.Errorf("couldn't retrieve latest checkpoint: %w", err)
 	}
-	// Parse the raw checkpoint into the log.Checkpoint format.
-	prev, err := w.parse(prevRaw, logID)
+	// Parse the raw retrieved checkpoint into the log.Checkpoint format.
+	prev, _, err := w.parse(prevRaw, logID)
 	if err != nil {
-		return 0, fmt.Errorf("couldn't parse stored checkpoint: %v", err)
+		return nil, fmt.Errorf("couldn't parse stored checkpoint: %v", err)
 	}
 	// Parse the compact range if we're using one.
 	var prevRange log.Proof
 	if logInfo.UseCompact {
 		if err := prevRange.Unmarshal(rangeRaw); err != nil {
-			return 0, fmt.Errorf("couldn't unmarshal proof: %v", err)
+			return nil, fmt.Errorf("couldn't unmarshal proof: %v", err)
 		}
 	}
 	if next.Size < prev.Size {
 		// Complain if prev is bigger than next.
-		return prev.Size, fmt.Errorf("cannot prove consistency backwards (%d < %d)", next.Size, prev.Size)
+		return prevRaw, status.Errorf(codes.Aborted, "cannot prove consistency backwards (%d < %d)", next.Size, prev.Size)
 	}
 	if next.Size == prev.Size {
 		if !bytes.Equal(next.Hash, prev.Hash) {
-			return prev.Size, fmt.Errorf("checkpoint for same size log with differing hash (got %x, have %x)", next.Hash, prev.Hash)
+			return prevRaw, status.Errorf(codes.Aborted, "checkpoint for same size log with differing hash (got %x, have %x)", next.Hash, prev.Hash)
```

Comment:
nit: is this more `FAILED_PRECONDITION`(neither this CP's nor any later CP on the same timeline is ever going to be consistent with the witness' one)

Also, it's maybe worth logging the full CP as evidence?

Reply:
+1. The difference is subtle at some level, but this isn't the kind of thing clients should just think they need to casually "retry at a higher level".

Reply:
Ah - I meant literally logging the CP (i.e. `glog.Warningf("Oh noes: %s", cpRaw)`) since it's potentially signed evidence of a log misbehaving.